datafusion/physical-plan/src/aggregates/row_hash.rs

```diff
@@ -465,19 +442,11 @@ impl Stream for GroupedHashAggregateStream {
 
                             // If the number of group values equals or exceeds the soft limit,
                             // emit all groups and switch to producing output
-                            if let Some(group_values_soft_limit) =
-                                self.group_values_soft_limit
-                            {
-                                if group_values_soft_limit <= self.group_values.len() {
-                                    timer.done();
-                                    if let Poll::Ready(Some(Err(e))) =
-                                        set_input_done_and_produce_output(&mut self)
-                                    {
-                                        return Poll::Ready(Some(Err(e)));
-                                    }
-                                    // make sure the exec_state just set is not overwritten below
-                                    break 'reading_input;
-                                }
+                            if self.hit_soft_group_limit() {
+                                timer.done();
+                                extract_ok!(self.set_input_done_and_produce_output());
```

Comment:
I think changing the signature of `set_input_done_and_produce_output` to return just a `Result<()>` simplifies handling and makes it clearer what can be returned